### PR TITLE
babe: treat epoch_authorship RPC method as unsafe

### PR DESCRIPTION
**Pull request by andresilva**
BABE's RPC API exposes a method to calculate all the slots that are allocated to the local keys in the current epoch. This information is sensitive and the method should not be exposed if the RPC is publicly addressable (like we do for authorship APIs).

Most of the changes in this PR are to allow the service builder to pass on a `DenyUnsafe` instance to the RPC extensions builder. Unfortunately the `DenyUnsafe` cannot be created at service instantiation time (since only when we try to actually bind http or ws interfaces do we know if it's unsafe), therefore we need to return a closure that builds the rpc extension. The resulting API (i.e. the new signature of `ServiceBuilder::with_rpc_extensions`) is a bit messy and complicated but doing this with a better API requires a refactor of the `ServiceBuilder` that I can't take on now.

polkadot companion: https://github.com/paritytech/polkadot/pull/1108

### CODE REVIEW

File: client/rpc-api/src/policy.rs [hunk 1]
```diff
@@ -25,37 +25,37 @@ use jsonrpc_core as rpc;
 /// Signifies whether a potentially unsafe RPC should be denied.
 #[derive(Clone, Copy, Debug)]
 pub enum DenyUnsafe {
-    /// Denies only potentially unsafe RPCs.
-    Yes,
-    /// Allows calling every RPCs.
-    No
+	/// Denies only potentially unsafe RPCs.
+	Yes,
+	/// Allows calling every RPCs.
```
Comment (DemiMarie-parity) [inline client/rpc-api/src/policy.rs:30]:
```suggestion
	/// Allows calling all RPCs.
```

File: client/rpc-api/src/policy.rs [hunk 2]
```diff
@@ -25,37 +25,37 @@ use jsonrpc_core as rpc;
...
 impl DenyUnsafe {
-    /// Returns `Ok(())` if the RPCs considered unsafe are safe to call,
-    /// otherwise returns `Err(UnsafeRpcError)`.
-    pub fn check_if_safe(self) -> Result<(), UnsafeRpcError> {
-        match self {
-            DenyUnsafe::Yes => Err(UnsafeRpcError),
-            DenyUnsafe::No => Ok(())
-        }
-    }
+	/// Returns `Ok(())` if the RPCs considered unsafe are safe to call,
```
Comment (DemiMarie-parity) [inline client/rpc-api/src/policy.rs:35]: Should the whitespace changes be a separate PR?

**Comment by andresilva** [comment]
The only unrelated whitespace change is in the `policy.rs` file, where the only changes I've made were reformatting with tabs instead of spaces (no content was changed).

File: client/service/src/builder.rs [hunk 3]
```diff
@@ -679,21 +679,27 @@ impl<TBl, TRtApi, TCl, TFchr, TSc, TImpQu, TFprb, TFpp, TExPool, TRpc, Backend>
...
-		let rpc_extensions = rpc_ext_builder(&self)?;
+		rpc_ext_builder: impl FnOnce(&Self) -> Result<URpcBuilder, Error>,
+	) -> Result<
+		ServiceBuilder<TBl, TRtApi, TCl, TFchr, TSc, TImpQu, TFprb, TFpp, TExPool, URpc, Backend>,
+		Error,
+	>
+	where
+		TSc: Clone,
+		TFchr: Clone,
+		URpcBuilder: Fn(sc_rpc::DenyUnsafe) -> URpc + Send + 'static,
```
Comment (tomusdrw) [inline client/service/src/builder.rs:700]: We could consider traitifying `Fn(sc_rpc::DenyUnsafe) -> URpc + Send + 'static` and implement that trait for both closures and concrete `URpc` type, so that we could maintain backward compatibility. Not sure if it's worth it though.

File: bin/node/cli/src/service.rs [hunk 4]
```diff
@@ -99,30 +99,46 @@ macro_rules! new_full_start {
...
+				let babe_config = babe_link.config().clone();
+				let shared_epoch_changes = babe_link.epoch_changes().clone();
+
+				let client = builder.client().clone();
+				let pool = builder.pool().clone();
+				let select_chain = builder.select_chain().cloned()
+					.expect("SelectChain is present for full services or set up failed; qed.");
+				let keystore = builder.keystore().clone();
+
+				Ok(move |deny_unsafe| -> RpcExtension {
```
Comment (tomusdrw) [inline bin/node/cli/src/service.rs:123]: Another way to avoid returning a closure would be to have `deny_unsafe` be a dynamic object that is lazy initialized when RPC is bound. That might require a `Mutex` and `thread_local` cache or some `unsafe` black magic to avoid huge performance costs on every RPC call though.

File: bin/node/cli/src/service.rs [hunk 5]
```diff
@@ -99,30 +99,46 @@ macro_rules! new_full_start {
 				import_setup = Some((block_import, grandpa_link, babe_link));
 				Ok(import_queue)
 			})?
-			.with_rpc_extensions(|builder| -> std::result::Result<RpcExtension, _> {
-				let babe_link = import_setup.as_ref().map(|s| &s.2)
-					.expect("BabeLink is present for full services or set up failed; qed.");
+			.with_rpc_extensions(|builder| {
```
Comment (tomusdrw) [inline bin/node/cli/src/service.rs:102]: We could expect a closure that get's a `RpcDependencies` object here instead of `Builder`, which would contain all required stuff and `deny_unsafe`. That would simplify the code a bit (no extra closure), but would make this method special in a way that it's not invoked right away like other `with_*` methods.

File: client/service/src/builder.rs (hunk 3 above)
Comment (andresilva) [reply client/service/src/builder.rs:700]: This might be a good idea actually, for example in the light client currently we don't care about `DenyUnsafe` for any rpc extensions. I guess right now anyone who isn't using BABE's RPC wouldn't care that much about it either.

**Comment by andresilva** [comment]
@tomusdrw I think the main problem is that unlike now, we might need to instantiate the rpc extensions with different parameters (currently we just clone the given extension when binding it to a server). It might be the case that we expose HTTP unsafely but not WS, in which case the RPC extension instances must have different `DenyUnsafe` values.

We also can't save a simple closure that takes `|builder, deny_unsafe|`, since the types of `builder` will change as we call the other extension methods. The only way to do this would be to somehow make `with_rpc_extensions` work differently as you suggested so that it is a terminal method, i.e. it builds the service.

File: bin/node/cli/src/service.rs (hunk 4 above)
Comment (andresilva) [reply bin/node/cli/src/service.rs:123]: But when we're building the service we don't know how many times the given extension will be initialized (http, ws?), and we'd need separate instances of this dynamic object for each instantiation I think.

Comment (tomusdrw) [reply bin/node/cli/src/service.rs:123]: Yeah, that's correct - we do need different parameter depending on the transport the extension will be used for, so this solution is a no-go.

File: client/service/src/builder.rs (hunk 3 above)
Comment (DemiMarie-parity) [reply client/service/src/builder.rs:700]: @tomusdrw how does one do that?  Every time I have tried to do that it fails with some obscure lifetime error.

**Comment by DemiMarie-parity** [comment]
@andresilva I will also add that IPC transports are virtually always safe to expose all RPCs on.

File: client/service/src/builder.rs (hunk 3 above)
Comment (tomusdrw) [reply client/service/src/builder.rs:700]: You mean this: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=59c9d81d289603e857e5a9dbb8970486 ?

**Comment by andresilva** [comment]
@tomusdrw I pushed a commit to traitify the RPC extension builder. It is still a breaking change as it requires wrapping the previous builder in some default builder implementation, the case were we take `DenyUnsafe` stays unchaged since we can convert from the closure to `RpcExtensionBuilder`. I have not added docs or updated the polkadot PR yet, let me know if you think this approach is good and we should move forward with it (or any other suggestions you might have). I am not sure whether this is worth it or not.

**Comment by andresilva** [comment]
I went with the suggestion of creating an `RpcExtensionBuilder` trait and in addition only use this trait in a new method `with_rpc_extensions_builder`, this makes this PR a non-breaking change as only RPC extensions that care about using `DenyUnsafe` need to use the builder. Updated the companion as well and added missing docs.
